internal/provider/resource_access_model.go

```diff
@@ -0,0 +1,269 @@
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	response, err := client.Users.ModelUserInfo(uuid)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if err := d.Set("model", id[0]); err != nil {
+		return diag.FromErr(err)
+	}
+	if err := d.Set("access", id[1]); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var users []string
+
+	for _, user := range stateUsers {
+		for _, modelUser := range response.ModelUserInfo {
+			if user == modelUser.UserName && string(modelUser.Access) == id[1] {
+				users = append(users, modelUser.UserName)
+			}
+		}
+	}
+
+	if err = d.Set("users", users); err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+// Updating the access model supports three cases
+// access and users both changed:
+// for missing users - revoke access
+// for changed users - apply new access
+// users changed:
+// for missing users - revoke access
+// for new users - apply access
+// access changed - apply new access
+func resourceAccessModelUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*juju.Client)
+
+	var diags diag.Diagnostics
+	anyChange := false
+
+	// items that could be changed
+	var newAccess string
+	var newUsersList []string
+	var missingUserList []string
+	var addedUserList []string
+
+	var err error
+
+	if d.HasChange("users") {
+		anyChange = true
+		oldUsers, newUsers := d.GetChange("users")
+		oldUsersInterface := oldUsers.([]interface{})
+		oldUsersList := make([]string, len(oldUsersInterface))
+		for i, v := range oldUsersInterface {
+			oldUsersList[i] = v.(string)
+		}
+		newUsersInterface := newUsers.([]interface{})
+		newUsersList = make([]string, len(newUsersInterface))
+		for i, v := range newUsersInterface {
+			newUsersList[i] = v.(string)
+		}
+		missingUserList = getMissingUsers(oldUsersList, newUsersList)
+		addedUserList = getAddedUsers(oldUsersList, newUsersList)
+	}
+
+	if !anyChange {
+		return diags
+	}
+
+	err = client.Models.UpdateAccessModel(juju.UpdateAccessModelInput{
+		Model:  d.Id(),
+		Grant:  addedUserList,
+		Revoke: missingUserList,
+		Access: newAccess,
+	})
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diags
+}
+
+func getMissingUsers(oldUsers, newUsers []string) []string {
+	var missing []string
+	for _, user := range oldUsers {
+		found := false
+		for _, newUser := range newUsers {
+			if user == newUser {
+				found = true
+				break
+			}
+		}
+		if !found {
+			missing = append(missing, user)
+		}
+	}
+	return missing
+}
+
+func getAddedUsers(oldUsers, newUsers []string) []string {
+	var added []string
+	for _, user := range newUsers {
+		found := false
+		for _, oldUser := range oldUsers {
+			if user == oldUser {
+				found = true
+				break
+			}
+		}
+		if !found {
+			added = append(added, user)
+		}
+	}
+	return added
+}
+
+// Juju refers to deletions as "destroy" so we call the Destroy function of our client here rather than delete
```

Comment:
Same comment about function docs.

Reply:
I adjusted the paragraph from existing resources that made use of the of "destroy" calls rather than "delete" and they start like that. I can change the paragraph if needed.

Reply:
Just for the of "Goness" I would just start the paragraph with a `resourceAccessModelDelete` for this comment to go into the godoc.

Reply:
Nice, I've added that.